### PR TITLE
Support OnDemandUpdate with multiple components at once.

### PR DESCRIPTION
**Pull request by goodov**
Changes in this PR:
1. The `CrxInstaller` now includes the overridable `IsBraveComponent()` function. This allows us to distinguish Brave/non-Brave components and let the updater to batch multiple update requests into a single request to `go-updater`. This is possible because Brave components are always managed by the `go-updater` and are never redirected to an upstream updater.
2. The `brave_component_updater::BraveOnDemandUpdater` is now the sole entry point for all Brave-specific "OnDemand" component updates.
3. The initialization sequence of `BraveOnDemandUpdater` has been altered to directly use `OnDemandUpdater` interface.

This PR will be followed with a few more to:
1. Cleanup `component_updater::BraveOnDemandUpdate`.
2. Replace `OnDemandUpdate` with conditional `Install` calls when a component is not installed. Currently, we perform unnecessary on demand updates at each startup. The `ComponentUpdateService` should manage most updates (and batch them) using the standard schedule.

Resolves https://github.com/brave/brave-browser/issues/37370

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: chromium_src/chrome/browser/component_updater/component_updater_utils.cc
Comment (goodov) [inline chromium_src/chrome/browser/component_updater/component_updater_utils.cc:12]: this function will be removed in a follow up PR. It's left for now to not overcomplicate the PR.

**Review by mherrmann — COMMENTED** [review]
Does upstream have the capability added by this PR already? If yes, is our reason for needing the extra code the special redirect logic in go-update? If the answer to both of these questions is yes, then maybe we could make changes in go-update and remove our customisations in brave-core.

Specifically: I know that go-update for example redirects update requests for Widevine to Google's server. This is due to a licensing restriction, which prevents us from serving Widevine. But what if go-update instead proxied the update request? So go-update would forward the request to Google's server, and then serve that as its response (for that one component). The response would still contain Google's binary download URL, so we would not be hosting Widevine. We would only be proxying the update request, much like a normal proxy.

File: components/brave_component_updater/browser/brave_on_demand_updater.cc
Comment (atuchin-m) [inline components/brave_component_updater/browser/brave_on_demand_updater.cc:52]: Could we add verification that  all `ids` are Brave components?
For chromium components better to call `OnDemandUpdate()` with a single id.

File: components/brave_component_updater/browser/brave_on_demand_updater.cc
Comment (atuchin-m) [inline components/brave_component_updater/browser/brave_on_demand_updater.cc:45]: Maybe `on_demand_updater_->OnDemandUpdate({id}, priority, std::move(callback));` to use the one entry point?

File: chromium_src/components/update_client/update_checker.cc
Comment (atuchin-m) [inline chromium_src/components/update_client/update_checker.cc:92]: Does it works well if `remaining_ids_` contains the on Chromium component?

File: components/brave_shields/core/browser/ad_block_component_service_manager.cc
Comment (atuchin-m) [inline components/brave_shields/core/browser/ad_block_component_service_manager.cc:290]: nit: remove `#include "base/barrier_callback.h"`

File: components/brave_shields/core/browser/ad_block_component_installer.cc
Comment (atuchin-m) [inline components/brave_shields/core/browser/ad_block_component_installer.cc:159]: @antonok-edm 
I wonder why `AdBlockComponentInstallerPolicy` doesn't implement `BraveComponentInstallerPolicy`?

File: components/playlist/browser/media_detector_component_installer.cc
Comment (atuchin-m) [inline components/playlist/browser/media_detector_component_installer.cc:140]: @goodov Could we check `IsBraveComponent() == true` in `BraveComponent::Register()` pipeline?

File: components/brave_component_updater/browser/brave_on_demand_updater.cc
Comment (goodov) [reply components/brave_component_updater/browser/brave_on_demand_updater.cc:52]: > Could we add verification that all `ids` are Brave components? For chromium components better to call `OnDemandUpdate()` with a single id.

Don't think we need this.

File: components/brave_shields/core/browser/ad_block_component_installer.cc
Comment (goodov) [reply components/brave_shields/core/browser/ad_block_component_installer.cc:159]: to me the `BraveComponent` and `BraveComponentInstallerPolicy` look like a legacy thing that should be cleaned up.

I'd like to keep this discussion out of the scope of this PR.

File: components/playlist/browser/media_detector_component_installer.cc
Comment (goodov) [reply components/playlist/browser/media_detector_component_installer.cc:140]: I'd like to avoid touching anything related to `BraveComponent` here.

**Comment by goodov** [comment]
PTAL everyone please. The PR is ready to be reviewed/merged.

> But what if go-update instead proxied the update request? So go-update would forward the request to Google's server, and then serve that as its response (for that one component). The response would still contain Google's binary download URL, so we would not be hosting Widevine. We would only be proxying the update request, much like a normal proxy. Perhaps that could avoid all the extra complexity?

The approach of modifying `go-updater` is reasonable and can be a second step to support batch update checks for all components, but it requires changes to the `go-updater` itself. This does look like a higher level separate task with some unknown variables at this stage.

**Review by kylehickinson — APPROVED** [review]
iOS++

**Review by supermassive — APPROVED** [review]
Wallet core lgtm

**Comment by mherrmann** [comment]
I respectfully repeat my concern. Upstream can update multiple components at once in the background, which imho is the important case. Unless batching on-demand component updates is really such high priority, I feel we should not pursue the approach in this PR but instead change go-update and get rid of all our customisations for serialising update requests. That's not a statement about the quality of this PR. I just feel that it is following a wrong path.

**Comment by mherrmann** [comment]
(that is, of course, if there are no blockers such as legal reasons that prevent us from going down that route.)

**Comment by goodov** [comment]
> Unless batching on-demand component updates is really such high priority

this is a requirement. AdBlocker components should be updated on a different schedule than other components. Upstream does not have a solution to configure the update schedule, so we had to run `OnDemanUpdate` on a custom schedule manually. To do this efficiently we need to support batching.

Please see the related PR https://github.com/brave/brave-core/pull/21868 and discussion https://github.com/brave/brave-core/pull/21868#discussion_r1475473908

**Comment by mherrmann** [comment]
@goodov if I understand the links you posted correctly, then the main motivation is that we need to update at least some components on a faster schedule (https://github.com/brave/brave-browser/issues/35164), which can also be controlled by Griffin.

At the moment, we perform many sequential update requests for the various components, at various times.

Wouldn't it be a lot easier to just reduce the normal background update check interval, and to revert our sequential update request logic in favor of an implementation in go-update that makes it proxy some bits of the update request to Google? Then we would only have one update request instead of the many we have now, and on a faster schedule.

**Comment by mherrmann** [comment]
> Can we actually proxy the batch-update check to Google?

The linked comment in this line talks about fingerprinting. I see that concern. Though I wonder how much better 20+ requests within one minute and I bet in 99.99999% coming from the same IP really are.

> we want to have AdBlock components updated at the very start.

Maybe I'm missing something, but I don't see how my proposal would change this. At the moment, Brave performs an update check for all components and extensions at browser start. I believe this comes from logic from upstream. But even if it didn't, we could kick off a background update check at start. I'd be surprised if that increased the latency until updated components are available by more than a few seconds - especially with one batch request instead of many single ones.

> Are we okay with increasing the check frequency for all components?

Besides the fingerprinting issue, I don't see what would be the harm. 

> This will also include installed extensions.

That's true. (I tested to confirm.)

> If we're switching AdBlock components to default, but more frequent update schedule, it might still not be enough

In which situation can it not be enough? We can set the background update interval as low as we want. The only scenario I can think of where background update checks are not enough is when components should be updated infrequently, but then when they are updated they need to be updated asap. I don't suspect that's the case.

**Comment by mherrmann** [comment]
> but AdBlock-specifics require OnDemandUpdate at the very start of the browser to update all its components as early as possible

@antonok-edm wrote the following in [Slack](https://bravesoftware.slack.com/archives/C3T9S9WKD/p1696086118642719?thread_ts=1695901547.243849&cid=C3T9S9WKD) about the motivation for updating ad block components early:

> I think that's a good idea too, but we have to be very careful about initial startup because if we delay too long users might see ads when they start the browser up

To me, this sounds like the motivation for updating ad components early is to make sure users don't see unwanted ads. But I don't think this requires OnDemandUpdate. On my machine, Chrome does a background update check for components within 60s of browser start. This means that I have the latest components after 61s. The 60s can be configured via `update_client::Configurator::InitialDelay()`. Instead of all the complexity we seem to be having around early on-demand update checks and batching, why don't we just set this to `0` or some other low value?

> AdBlock components update on the start must be in a single batch to properly handle the re-building of the AdBlock engine only when ALL AdBlock-related components are downloaded. 

With `InitialDelay` set to a low value, the components would naturally be updated in a single batch on browser start. Just like this PR, such an implementation could also serve as a foundation for re-building the ad block engine only once.

As a related note, some of the ad block performance problems came from code running on the UI thread. Maybe upstream has or will have more performance-friendly logic for background update checks vs. on-demand update checks that we could benefit from. Using an on-demand update check when a background update check would probably be enough is not a good idea in general in my opinion.

> but it's a task that aims to optimize the general component updater flow, not the task that solves the AdBlock components update problem. Please create a separate issue to work on this, I would be happy if we remove SequentialUpdateChecker entirely.

The task is to update ad block components in a single batch. There is an easy, built-in mechanism for doing so. I feel we should use it. I have no doubt that the quality of this PR is excellent. But it seems to me that this is leading us down a wrong path. If we go ahead, future development work will build on the APIs introduced here. But I have the impression that we are working against upstream's design here, and I believe our lives will be easier if we instead work with it.

**Comment by goodov** [comment]
> Instead of all the complexity we seem to be having around early on-demand update checks and batching, why don't we just set this to 0 or some other low value?
<...>
Using an on-demand update check when a background update check would probably be enough is not a good idea in general in my opinion.

1. This will trigger an update of all components, including extensions, which will not finish until all components are updated.
2. Each component update will trigger `OnComponentReady` in an undefined order and time.

> But I have the impression that we are working against upstream's design here

Usptream does not have the scenario of interlinked components.

> The task is to update ad block components in a single batch.

... and to get the completion status of the update in a single callback.

> There is an easy, built-in mechanism for doing so.

There is no public API in updater interfaces to request the update of specific components and get a single callback when all requested components are updated.

**Comment by mherrmann** [comment]
Alright, I could make further points in response to your comments, but it seems your mind is made up and the discussion will not go anywhere. Feel free to merge without my approval. 👍

**Review by Brandon-T — APPROVED** [review]

File: components/brave_shields/core/browser/ad_block_component_installer.cc
Comment (antonok-edm) [reply components/brave_shields/core/browser/ad_block_component_installer.cc:159]: > to me the `BraveComponent` and `BraveComponentInstallerPolicy` look like a legacy thing that should be cleaned up.

Yep, this code is from before my time at Brave and I have no attachment to it. But agreed that it doesn't need to be handled here.

**Review by antonok-edm — APPROVED** [review]
thanks @goodov! Your assessment of the situation looks accurate to me - there's some chance we can re-evaluate the approach after a generalized solution to https://github.com/brave/brave-core/pull/22188 is available for all the adblock-related components, but it's not realistic in the short term.

**Comment by goodov** [comment]
@brave/sec-team do we need a security review here? There was a fingerprinting-related question that seems to be answered https://github.com/brave/brave-browser/issues/37370#issuecomment-2043609479

If we do need a sec review here, I will create it. Otherwise, please review this PR and remove `needs-security-review` label.

**Comment by fmarier** [comment]
The needs-sec-review label was added because of the two uses of `base::Unretained` which you have already reviewed and resolved. No need to file a full sec review.

**Pull request merged**
